myco_api_client/matching_algorithms/preferred_partners_algorithm.py

```diff
@@ -0,0 +1,173 @@
+            bid_requirement: Dict, offer_requirement: Dict) -> bool:
+        """
+        Check if we can match offer & bid taking their selected requirements into consideration.
+        """
+
+        offer_required_energy, offer_required_clearing_rate = (
+            cls._get_required_energy_and_rate_from_order(offer, offer_requirement))
+
+        bid_required_energy, bid_required_clearing_rate = (
+            cls._get_required_energy_and_rate_from_order(bid, bid_requirement))
+        if bid_required_clearing_rate < offer_required_clearing_rate:
+            return False
+        if not (RequirementsSatisfiedChecker.is_bid_requirement_satisfied(
+                bid=bid, offer=offer,
+                selected_energy=min(offer_required_energy, bid_required_energy),
+                clearing_rate=bid_required_clearing_rate,
+                bid_requirement=bid_requirement) and
+                RequirementsSatisfiedChecker.is_offer_requirement_satisfied(
+                    bid=bid, offer=offer,
+                    selected_energy=min(
+                        offer_required_energy, bid_required_energy),
+                    clearing_rate=bid_required_clearing_rate,
+                    offer_requirement=offer_requirement)):
+            return False
+        return True
+
+    @staticmethod
+    def _get_actor_to_offers_mapping(
+            offers: List[Offer.serializable_dict]
+    ) -> Dict[str, List[Offer.serializable_dict]]:
+        """Map seller ids/origin ids to their offers list."""
+        mapping = {}
+        for offer in offers:
+            if offer["seller_id"]:
+                if offer["seller_id"] not in mapping:
+                    mapping[offer["seller_id"]] = []
+                mapping[offer["seller_id"]].append(offer)
+            if (offer["seller_origin_id"]
+                    and offer["seller_origin_id"] != offer["seller_id"]):
+                if offer["seller_origin_id"] not in mapping:
+                    mapping[offer["seller_origin_id"]] = []
+                mapping[offer["seller_origin_id"]].append(offer)
+        return mapping
+
+    @classmethod
+    def _get_required_energy_and_rate_from_order(
+            cls, order: BaseBidOffer.serializable_dict,
+            order_requirement: Dict) -> Tuple[float, float]:
+        """Determine the energy and clearing rate based on an order + its requirement.
+
+        A bid or offer can have energy and clearing rate attributes on both the instance
+        and as a special requirement.
+        The values in requirement have higher priority in selecting the energy and rate.
+
+        Args:
+            order: a serialized offer or bid structures
+            order_requirement: specified requirement dictionary for the order
+        """
+        order_required_energy = order_requirement.get("energy") or order["energy"]
+        order_required_clearing_rate = (
+                order_requirement.get("price") or order["energy_rate"])
```

Comment:
```suggestion
                order_requirement.get("energy_rate") or order["energy_rate"])
```

Reply:
Thanks for catching that!
We don't actually have a `energy_rate` requirement, we only have a `price` requirement as in the above line, however I indeed have to make it into an energy rate

Reply:
Thanks, test are failing after last commit. Can you please look into it

Reply:
They were due to the bug discussed on slack, they passed now